posix: Map CLOCK_REALTIME and CLOCK_MONOTONIC to Zephyr clock IDs

M-Moawad opened the pull request:
Some toolchains may define **CLOCK_REALTIME and 
CLOCK_MONOTONIC** in their libc headers
with values that differ from Zephyr's internal 
**SYS_CLOCK_REALTIME and SYS_CLOCK_MONOTONIC**.

To ensure consistent behavior across all toolchains, Introduce a helper function to map
**CLOCK_REALTIME and CLOCK_MONOTONIC** to Zephyr's internal
clock IDs **(SYS_CLOCK_REALTIME and SYS_CLOCK_MONOTONIC)**.

This prevents mismatched clock IDs being passed to  the kernel, avoiding invalid clockid errors when using  functions like **clock_gettime()**.

Fixing issue: https://github.com/zephyrproject-rtos/zephyr/issues/93073

cfriedt reviewed (approved):
Looks good to me.

cfriedt commented:
@M-Moawad - it looks like your commit message needs to be reformatted.

https://github.com/zephyrproject-rtos/zephyr/actions/runs/16251745415/job/45886336337#step:10:1

Use `./scripts/ci/check_compliance.py -c main..` prior to pushing for checking locally.

tagunil commented:
I'm not sure this will work long term because if you have a toolchain with a C library built against the old values, and then you redefine them to the Zephyr ones, you will have hard-to-debug issues calling clock_gettime and other functions from the C library dependent of whether you have this header included or not. Although AFAIU this will also depend on whether TC_PROVIDES_POSIX_TIMERS is set for the toolchain. @cfriedt, what do you think, where all of this should be going?

cfriedt commented:
@M-Moawad @tagunil - can you check if this works with ARC MWDT?

```diff
diff --git a/include/zephyr/sys/clock.h b/include/zephyr/sys/clock.h
index 47fbfa3cf47..00e0a9a1ade 100644
--- a/include/zephyr/sys/clock.h
+++ b/include/zephyr/sys/clock.h
@@ -378,6 +378,9 @@ static inline bool sys_timepoint_expired(k_timepoint_t timepoint)
 /** @cond INTERNAL_HIDDEN */
 /* forward declaration as workaround for time.h */
 struct timespec;
+
+/* Convert a POSIX clock (cast to int) to a sys_clock identifier */
+int sys_clock_from_clockid(int clock_id);
 /** @endcond INTERNAL_HIDDEN */
 
 /**
diff --git a/lib/os/clock.c b/lib/os/clock.c
index 98867ea3675..ca0973cb505 100644
--- a/lib/os/clock.c
+++ b/lib/os/clock.c
@@ -50,6 +50,22 @@ static void timespec_from_ticks(uint64_t ticks, struct timespec *ts)
        };
 }
 
+int sys_clock_from_clockid(int clock_id)
+{
+       switch (clock_id) {
+#if defined(_POSIX_C_SOURCE)
+       case CLOCK_REALTIME:
+               return SYS_CLOCK_REALTIME;
+#endif
+#if defined(_POSIX_MONOTONIC_CLOCK)
+       case CLOCK_MONOTONIC:
+               return SYS_CLOCK_MONOTONIC;
+#endif
+       default:
+               return -EINVAL;
+       }
+}
+
 int sys_clock_gettime(int clock_id, struct timespec *ts)
 {
        if (!is_valid_clock_id(clock_id)) {
diff --git a/lib/posix/options/clock.c b/lib/posix/options/clock.c
index 239dadb6926..466974705ac 100644
--- a/lib/posix/options/clock.c
+++ b/lib/posix/options/clock.c
@@ -18,7 +18,7 @@ int clock_gettime(clockid_t clock_id, struct timespec *ts)
 {
        int ret;
 
-       ret = sys_clock_gettime((int)clock_id, ts);
+       ret = sys_clock_gettime(sys_clock_from_clockid((int)clock_id), ts);
        if (ret < 0) {
                errno = -ret;
                return -1;
@@ -61,7 +61,7 @@ int clock_settime(clockid_t clock_id, const struct timespec *tp)
 {
        int ret;
 
-       ret = sys_clock_settime((int)clock_id, tp);
+       ret = sys_clock_settime(sys_clock_from_clockid((int)clock_id), tp);
        if (ret < 0) {
                errno = -ret;
                return -1;
diff --git a/lib/posix/options/clock_selection.c b/lib/posix/options/clock_selection.c
index c67d48ff025..5717221b596 100644
--- a/lib/posix/options/clock_selection.c
+++ b/lib/posix/options/clock_selection.c
@@ -26,7 +26,7 @@ int clock_nanosleep(clockid_t clock_id, int flags, const struct timespec *rqtp,
                return -1;
        }
 
-       ret = sys_clock_nanosleep((int)clock_id, flags, rqtp, rmtp);
+       ret = sys_clock_nanosleep(sys_clock_from_clockid((int)clock_id), flags, rqtp, rmtp);
        if (ret < 0) {
                errno = -ret;
                return -1;
```

I think that should handle all use cases, but please correct me if there are any call sites missing.

```shell
$ twister --retry-failed 3 -T tests/posix/timers/
...
INFO    - Total complete:   40/  40  100%  built (not run):    0, filtered:  145, failed:    0, error:    0
INFO    - 4 test scenarios (184 configurations) selected, 145 configurations filtered (144 by static filter, 1 at runtime).
INFO    - 39 of 39 executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 164.44 seconds.
INFO    - 624 of 624 executed test cases passed (100.00%) on 10 out of total 1115 platforms (0.90%).
INFO    - 39 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

cfriedt reviewed (changes requested):
As mentioned in the linked comment, undefining CLOCK_REALTIME and CLOCK_MONOTONIC might not address the issue effectively. 
https://github.com/zephyrproject-rtos/zephyr/pull/93075/#issuecomment-3069697962

It's probably best to create a function that maps `CLOCK_REALTIME` and `CLOCK_MONOTONIC` to `SYS_CLOCK_REALTIME` and `SYS_CLOCK_MONOTONIC`.

M-Moawad committed:
posix: Map CLOCK_REALTIME and CLOCK_MONOTONIC to Zephyr clock IDs

Some toolchains may define CLOCK_REALTIME and
CLOCK_MONOTONIC in their libc headers
with values that differ from Zephyr's internal
SYS_CLOCK_REALTIME and SYS_CLOCK_MONOTONIC.

To ensure consistent behavior across all boards and
toolchains, Introduce a helper function to map
CLOCK_REALTIME and CLOCK_MONOTONIC to Zephyr's internal
clock IDs (SYS_CLOCK_REALTIME and SYS_CLOCK_MONOTONIC).

This prevents mismatched clock IDs being passed to
the kernel, avoiding invalid clockid errors when using
functions like clock_gettime().

Signed-off-by: Mohamed Moawad <moawad@synopsys.com>

M-Moawad commented:
> @M-Moawad @tagunil - can you check if this works with ARC MWDT? It might be the least amount of work in this case.

I tested your patch locally with ARC MWDT toolchain and it passed, so I’ve applied it—it’s definitely a cleaner solution. I’ve also triggered a pipeline job to run the full test suite, and I’ll let you know the results soon.

cfriedt reviewed (approved):
Thanks @M-Moawad - looks good to me

M-Moawad commented:
@tagunil Please review the current change

tagunil reviewed (approved):
I have in mind something more than this for MWDT toolchain specifically, but for the issue at hand that looks like a valid solution. @M-Moawad , could you please also update PR description to reflect the new approach?

fabiobaltieri commented on lib/os/clock.c:
```diff
@@ -50,6 +50,22 @@ static void timespec_from_ticks(uint64_t ticks, struct timespec *ts)
+int sys_clock_from_clockid(int clock_id)
+{
+	switch (clock_id) {
+#if defined(CLOCK_REALTIME) || defined(_POSIX_C_SOURCE)
+	case (int)CLOCK_REALTIME:
+		return SYS_CLOCK_REALTIME;
+#endif
+#if defined(CLOCK_MONOTONIC) || defined(_POSIX_MONOTONIC_CLOCK)
+	case (int)CLOCK_MONOTONIC:
+		return SYS_CLOCK_MONOTONIC;
+#endif
```
hey are this right? traced a regression in a downstream application down to this neither of the two options are complied in, is this missing an `#include <zephyr/posix/time.h>` possibly?

M-Moawad replied:
You’re right , but I have an inquiry: which libc are you using, and does it actually define CLOCK_REALTIME and CLOCK_MONOTONIC?
From my side, I see the include path in lib/os/clock.c as:

`lib/os/clock.c → zephyr/sys/timeutil.h → zephyr/posix/time.h → libc <time.h>`

cfriedt replied:
@fabiobaltieri - can you provide a repro recipe?

We shouldn't necessarily need to include `posix/time.h` if the libc `time.h` is included assuming the libc `time.h` defines either of these two.

It might require a `#define _POSIX_C_SOURCE` before including `time.h` though.

Also, see if #95226 helps if you like.

fabiobaltieri replied:
Hey Chris, yeah looks like #95226 fixes it, let's get on with it. Thanks!
